amqVivaceShortcuts.user.js

```diff
@@ -317,14 +334,53 @@ const optimizedShortcuts = (targets) => {
         minPos = pos;
         bestSubstring = substring;
       }
-
-      // If a perfect shortcut is found, append it to the results
-      if (pos === 0) {
-        shortcuts.push(substring);
+      substring = substring + "↓".repeat(pos);
+      if (substring.length > MAX_SUBSTRING_LENGTH) {
+        continue;
+      }
+      //If the shortcut found has a substring that's already in shortcuts list don't add it. Otherwise, do add it.
+      let superStringQ = false;
+      if (!FULL_SHORTCUTS) {
+        for (const currentShortcut of shortcuts.concat(altShortcuts)) {
+          if (substring.length < currentShortcut.length) {
+            continue;
+          }
+          let i = 0;
+          let j = 0;
+          while (i < substring.length && j < currentShortcut.length) {
+            if (currentShortcut[j] == substring[i]) {
+              j++;
+            }
+            i++;
+          }
+          if (j >= currentShortcut.length) {
+            superStringQ = true;
+            break;
+          }
+        }
+      }
+      //if (shortcuts.find(shortcut => substring.includes(shortcut)) == undefined || FULL_SHORTCUTS){
+      if (!superStringQ) {
+        if (pos == 0) {
+          shortcuts.push(substring);
+        } else {
+          altShortcuts.push(substring);
+        }
       }
     }
   }
-
+  while (altShortcuts.length > 0 && shortcuts.length < NUM_SHORTCUTS) {
+    let temp = [];
+    for (const altSubString of altShortcuts) {
+      if (altSubString.length == currentLength) {
+        shortcuts.push(altSubString);
+      } else {
+        temp.push(altSubString);
+      }
+    }
+    altShortcuts = temp;
+    currentLength += 1;
+  }
   return shortcuts.length ? shortcuts : [bestSubstring];
```

Comment:
I'd suggest to use a sort here, so that it's easier to understand that you are taking all the shortest `altShortcuts` in order.

```suggestion

  // If not enough shortcuts were found, try to fill with the alternative shortcuts to reach at least NUM_SHORTCUTS. Include all the 
  if (altShortcuts.length > 0 && shortcuts.length < NUM_SHORTCUTS) {
    altShortcuts.sort((a, b) => a.length - b.length);

    // Take at least enough to fill up to NUM_SHORTCUTS
    const neededCount = Math.min(
      NUM_SHORTCUTS - shortcuts.length,
      altShortcuts.length
    );

    const maxAltLength = altShortcuts[neededCount - 1].length;

    // Take all the altShortcuts with the same length as the longest that is needed 
    shortcuts = shortcuts.concat(altShortcuts.filter(s => s.length <= maxAltLength));
  }

  return shortcuts.length ? shortcuts : [bestSubstring];
```